rfc/20250728-opentofu-providers/03-provider-sdk.md

```diff
@@ -0,0 +1,618 @@
+
+**3. Language-Idiomatic Design**
+Each SDK follows the conventions and patterns of its target language ecosystem, making provider development feel natural to developers already familiar with that language.
+
+**4. Progressive Disclosure**
+Simple use cases require minimal code, while complex scenarios remain possible through advanced SDK features and escape hatches.
+
+**5. Automatic Features**
+The SDK automatically should provide managed resources, data sources, documentation generation, capability negotiation, and protocol handling, etc. without requiring developer intervention.
+
+## SDK Architecture
+
+### Transport Abstraction
+
+All SDKs use a transport abstraction that handles the underlying protocol communication. The initial implementation focuses on stdio transport, with future support for other transports:
+
+```typescript
+// TypeScript example
+import { Provider, StdioTransport } from '@opentofu/provider-sdk';
+
+const provider = new Provider({ name: "custom", version: "1.0.0" });
+new StdioTransport().connect(provider);
+```
+
+### Protocol Translation Layer
+
+The SDK acts as a translation layer between the developer's simple methods and something that could fulfil the `providers.interface` specification. See [02-provider-client-library.md](./02-provider-client-library.md) for more details.
+
+For example a developer should be able to write the following:
+```typescript
+provider.resource("s3_bucket", {
+  ...
+  methods: {
+    async create(config) {
+      const result = await my_s3_client.createBucket(config);
+      return { id: result.id, state: { ...config, id: result.id } };
+    }
+  }
+}
+```
+
+And the SDK translates the calls to/from such functions:
+- `GetProviderSchema` responses with proper schema definitions
+- `PlanResourceChange` handling with unknown value management
+- `ApplyResourceChange` execution with error handling
+- State persistence and retrieval logic
+
+## Multi-Language Implementation
+
+### TypeScript SDK
+
+> [!NOTE]
+> The proposal for typescript may seem more comlpete as it is what I am most familiar with. Other language examples are shown to define the pattern of implementation as an example, and not as something set in stone.
+
+The TypeScript SDK could leverage Zod for schema validation and provides a simple experience of defining items on the provider. For example:
+
+```typescript
+import { z } from 'zod';
+import { Provider, StdioTransport } from '@opentofu/provider-sdk';
+
+const provider = new Provider({
+  name: "my-custom-aws-s3",
+  version: "0.1.0",
+});
+
+// Schema-first resource definition
+const s3BucketSchema = z.object({
+  bucket: z.string(),
+  region: z.string().default("us-east-1"),
+  versioning: z.boolean().default(false),
+  tags: z.record(z.string()).optional(),
+  // Computed fields
+  arn: z.string().optional(),
+  id: z.string().optional(),
+});
+
+provider.resource("s3_bucket", {
+  schema: s3BucketSchema,
+  methods: {
+    async read(id, config) {
```

Comment:
I'd expect that a TypeScript developer would prefer to have an explicit type constraint on `config` here, so that e.g. their editor will do code completion of what properties are expected in this object type.

This is really a general TypeScript/zod question rather than OpenTofu-specific, but is there some idiomatic way to either generate a compile-time TypeScript interface from a runtime zod schema or vice-versa?